chromadb/auth/basic_authn/__init__.py

```diff
@@ -0,0 +1,130 @@
+import base64
+import importlib
+import logging
+
+from fastapi import HTTPException
+from overrides import override
+from pydantic import SecretStr
+
+from chromadb.auth import (
+    UserIdentity,
+    ServerAuthenticationProvider,
+    ClientAuthProvider,
+    ClientAuthHeaders,
+)
+from chromadb.config import System
+from chromadb.telemetry.opentelemetry import (
+    OpenTelemetryGranularity,
+    trace_method,
+)
+from starlette.datastructures import Headers
+from typing import Dict
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["BasicAuthenticationServerProvider", "BasicAuthClientProvider"]
+
+
+class BasicAuthClientProvider(ClientAuthProvider):
+    """
+    Client auth provider for basic auth. The credentials are passed as a
+    base64-encoded string in the Authorization header prepended with "Basic ".
+    """
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self._settings = system.settings
+        system.settings.require("chroma_client_auth_credentials")
+        self._creds = SecretStr(
+            str(system.settings.chroma_client_auth_credentials)
+        )
+
+    @override
+    def authenticate(self) -> ClientAuthHeaders:
+        # We do this goofy encode-decode dance to make the type checker happy.
+        encoded = base64.b64encode(
+            f"{self._creds.get_secret_value()}".encode("utf-8")
+        ).decode(
+            "utf-8"
+        )
+        return {
+            "Authorization": SecretStr(
+                f"Basic {encoded}"
+            ),
+        }
+
+
+class BasicAuthenticationServerProvider(ServerAuthenticationProvider):
+    """
+    Server auth provider for basic auth. The credentials are read from
+    `chroma_server_authn_credentials_file` and each line must be in the format
+    <username>:<bcrypt passwd>.
+
+    Expects tokens to be passed as a base64-encoded string in the Authorization
+    header prepended with "Basic".
+    """
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self._settings = system.settings
+
+        try:
+            # We need this to check passwords
+            self.bc = importlib.import_module("bcrypt")
```

Comment:
seems like bcrypt is in our requirements - https://github.com/chroma-core/chroma/blob/main/requirements.txt

and our toml https://github.com/chroma-core/chroma/blob/f44d9e82d2c1ff4348f0ddd8ae32597f3bccf057/pyproject.toml#L39

Why do we dynamically import the module like this?

Reply:
Great question. Removed.